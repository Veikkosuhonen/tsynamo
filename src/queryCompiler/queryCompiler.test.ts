import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";
import { DDB, TEST_DATA } from "../../test/testFixture";
import { getDDBClientFor, startDDBTestContainer } from "../../test/testUtil";
import { Tsynamo } from "./../index";

describe("QueryQueryBuilder", () => {
  let tsynamoClient: Tsynamo<DDB>;

  beforeAll(async () => {
    tsynamoClient = new Tsynamo<DDB>({
      ddbClient: DynamoDBDocumentClient.from(new DynamoDBClient({})),
    });
  });

  it("queryQueryBuilder can be compiled", () => {
    const data = tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("someBoolean", "=", true)
      .orFilterExpression("somethingElse", "BETWEEN", 9, 10)
      .attributes(["userId"])
      .compile();

    expect(data).toMatchSnapshot();
  });

  it("putItemQueryBuilder can be compiled", () => {
    const data = tsynamoClient
      .putItem("myTable")
      .item({
        userId: "333",
        dataTimestamp: 212,
        tags: ["cats"],
      })
      .compile();

    expect(data).toMatchSnapshot();
  });

  it("getItemQueryBuilder can be compiled", () => {
    const data = tsynamoClient
      .getItem("myTable")
      .keys({
        userId: TEST_DATA[1].userId,
        dataTimestamp: TEST_DATA[1].dataTimestamp,
      })
      .compile();

    expect(data).toMatchSnapshot();
  });

  it("deleteItemQueryBuilder can be compiled", () => {
    const data = tsynamoClient
      .deleteItem("myTable")
      .keys({
        userId: "1",
        dataTimestamp: 2,
      })
      .returnValues("ALL_OLD")
      .compile();

    expect(data).toMatchSnapshot();
  });

  it("updateItemQueryBuilder can be compiled", () => {
    const data = tsynamoClient
      .updateItem("myTable")
      .keys({
        userId: "1",
        dataTimestamp: 2,
      })
      .add("someSet", new Set(["1", "2"]))
      .set("nested.nestedBoolean", "=", true)
      .remove("someBoolean")
      .delete("nested.nestedSet", new Set(["a"]))
      .compile();

    expect(data).toMatchSnapshot();
  });
});
