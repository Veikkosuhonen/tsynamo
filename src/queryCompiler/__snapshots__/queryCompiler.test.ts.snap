// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QueryQueryBuilder > deleteItemQueryBuilder can be compiled 1`] = `
DeleteCommand {
  "clientCommand": DeleteItemCommand {
    "deserialize": [Function],
    "input": {
      "ConditionExpression": undefined,
      "ExpressionAttributeNames": undefined,
      "ExpressionAttributeValues": undefined,
      "Key": {
        "dataTimestamp": 2,
        "userId": "1",
      },
      "ReturnValues": "ALL_OLD",
      "ReturnValuesOnConditionCheckFailure": undefined,
      "TableName": "myTable",
    },
    "middlewareStack": {
      "add": [Function],
      "addRelativeTo": [Function],
      "applyToStack": [Function],
      "clone": [Function],
      "concat": [Function],
      "identify": [Function],
      "identifyOnResolve": [Function],
      "remove": [Function],
      "removeByTag": [Function],
      "resolve": [Function],
      "use": [Function],
    },
    "serialize": [Function],
  },
  "input": {
    "ConditionExpression": undefined,
    "ExpressionAttributeNames": undefined,
    "ExpressionAttributeValues": undefined,
    "Key": {
      "dataTimestamp": 2,
      "userId": "1",
    },
    "ReturnValues": "ALL_OLD",
    "ReturnValuesOnConditionCheckFailure": undefined,
    "TableName": "myTable",
  },
  "inputKeyNodes": {
    "Expected": {
      "*": {
        "AttributeValueList": [],
        "Value": null,
      },
    },
    "ExpressionAttributeValues": {},
    "Key": {},
  },
  "middlewareStack": {
    "add": [Function],
    "addRelativeTo": [Function],
    "applyToStack": [Function],
    "clone": [Function],
    "concat": [Function],
    "identify": [Function],
    "identifyOnResolve": [Function],
    "remove": [Function],
    "removeByTag": [Function],
    "resolve": [Function],
    "use": [Function],
  },
  "outputKeyNodes": {
    "Attributes": {},
    "ItemCollectionMetrics": {
      "ItemCollectionKey": {},
    },
  },
}
`;

exports[`QueryQueryBuilder > getItemQueryBuilder can be compiled 1`] = `
GetCommand {
  "clientCommand": GetItemCommand {
    "deserialize": [Function],
    "input": {
      "ConsistentRead": undefined,
      "ExpressionAttributeNames": undefined,
      "Key": {
        "dataTimestamp": 333,
        "userId": "321",
      },
      "ProjectionExpression": undefined,
      "TableName": "myTable",
    },
    "middlewareStack": {
      "add": [Function],
      "addRelativeTo": [Function],
      "applyToStack": [Function],
      "clone": [Function],
      "concat": [Function],
      "identify": [Function],
      "identifyOnResolve": [Function],
      "remove": [Function],
      "removeByTag": [Function],
      "resolve": [Function],
      "use": [Function],
    },
    "serialize": [Function],
  },
  "input": {
    "ConsistentRead": undefined,
    "ExpressionAttributeNames": undefined,
    "Key": {
      "dataTimestamp": 333,
      "userId": "321",
    },
    "ProjectionExpression": undefined,
    "TableName": "myTable",
  },
  "inputKeyNodes": {
    "Key": {},
  },
  "middlewareStack": {
    "add": [Function],
    "addRelativeTo": [Function],
    "applyToStack": [Function],
    "clone": [Function],
    "concat": [Function],
    "identify": [Function],
    "identifyOnResolve": [Function],
    "remove": [Function],
    "removeByTag": [Function],
    "resolve": [Function],
    "use": [Function],
  },
  "outputKeyNodes": {
    "Item": {},
  },
}
`;

exports[`QueryQueryBuilder > putItemQueryBuilder can be compiled 1`] = `
PutCommand {
  "clientCommand": PutItemCommand {
    "deserialize": [Function],
    "input": {
      "ConditionExpression": undefined,
      "ExpressionAttributeNames": undefined,
      "ExpressionAttributeValues": undefined,
      "Item": {
        "dataTimestamp": 212,
        "tags": [
          "cats",
        ],
        "userId": "333",
      },
      "ReturnValues": undefined,
      "TableName": "myTable",
    },
    "middlewareStack": {
      "add": [Function],
      "addRelativeTo": [Function],
      "applyToStack": [Function],
      "clone": [Function],
      "concat": [Function],
      "identify": [Function],
      "identifyOnResolve": [Function],
      "remove": [Function],
      "removeByTag": [Function],
      "resolve": [Function],
      "use": [Function],
    },
    "serialize": [Function],
  },
  "input": {
    "ConditionExpression": undefined,
    "ExpressionAttributeNames": undefined,
    "ExpressionAttributeValues": undefined,
    "Item": {
      "dataTimestamp": 212,
      "tags": [
        "cats",
      ],
      "userId": "333",
    },
    "ReturnValues": undefined,
    "TableName": "myTable",
  },
  "inputKeyNodes": {
    "Expected": {
      "*": {
        "AttributeValueList": [],
        "Value": null,
      },
    },
    "ExpressionAttributeValues": {},
    "Item": {},
  },
  "middlewareStack": {
    "add": [Function],
    "addRelativeTo": [Function],
    "applyToStack": [Function],
    "clone": [Function],
    "concat": [Function],
    "identify": [Function],
    "identifyOnResolve": [Function],
    "remove": [Function],
    "removeByTag": [Function],
    "resolve": [Function],
    "use": [Function],
  },
  "outputKeyNodes": {
    "Attributes": {},
    "ItemCollectionMetrics": {
      "ItemCollectionKey": {},
    },
  },
}
`;

exports[`QueryQueryBuilder > queryQueryBuilder can be compiled 1`] = `
QueryCommand {
  "clientCommand": QueryCommand {
    "deserialize": [Function],
    "input": {
      "ConsistentRead": undefined,
      "ExpressionAttributeNames": {
        "#someBoolean": "someBoolean",
        "#somethingElse": "somethingElse",
        "#userId": "userId",
      },
      "ExpressionAttributeValues": {
        ":filterExpressionValue0": true,
        ":filterExpressionValue1left": 9,
        ":filterExpressionValue1right": 10,
        ":keyConditionValue0": "123",
      },
      "FilterExpression": "#someBoolean = :filterExpressionValue0 OR #somethingElse BETWEEN :filterExpressionValue1left AND :filterExpressionValue1right",
      "KeyConditionExpression": "#userId = :keyConditionValue0",
      "Limit": undefined,
      "ProjectionExpression": "#userId",
      "ScanIndexForward": undefined,
      "TableName": "myTable",
    },
    "middlewareStack": {
      "add": [Function],
      "addRelativeTo": [Function],
      "applyToStack": [Function],
      "clone": [Function],
      "concat": [Function],
      "identify": [Function],
      "identifyOnResolve": [Function],
      "remove": [Function],
      "removeByTag": [Function],
      "resolve": [Function],
      "use": [Function],
    },
    "serialize": [Function],
  },
  "input": {
    "ConsistentRead": undefined,
    "ExpressionAttributeNames": {
      "#someBoolean": "someBoolean",
      "#somethingElse": "somethingElse",
      "#userId": "userId",
    },
    "ExpressionAttributeValues": {
      ":filterExpressionValue0": true,
      ":filterExpressionValue1left": 9,
      ":filterExpressionValue1right": 10,
      ":keyConditionValue0": "123",
    },
    "FilterExpression": "#someBoolean = :filterExpressionValue0 OR #somethingElse BETWEEN :filterExpressionValue1left AND :filterExpressionValue1right",
    "KeyConditionExpression": "#userId = :keyConditionValue0",
    "Limit": undefined,
    "ProjectionExpression": "#userId",
    "ScanIndexForward": undefined,
    "TableName": "myTable",
  },
  "inputKeyNodes": {
    "ExclusiveStartKey": {},
    "ExpressionAttributeValues": {},
    "KeyConditions": {
      "*": {
        "AttributeValueList": [],
      },
    },
    "QueryFilter": {
      "*": {
        "AttributeValueList": [],
      },
    },
  },
  "middlewareStack": {
    "add": [Function],
    "addRelativeTo": [Function],
    "applyToStack": [Function],
    "clone": [Function],
    "concat": [Function],
    "identify": [Function],
    "identifyOnResolve": [Function],
    "remove": [Function],
    "removeByTag": [Function],
    "resolve": [Function],
    "use": [Function],
  },
  "outputKeyNodes": {
    "Items": {
      "*": {},
    },
    "LastEvaluatedKey": {},
  },
}
`;

exports[`QueryQueryBuilder > updateItemQueryBuilder can be compiled 1`] = `
UpdateCommand {
  "clientCommand": UpdateItemCommand {
    "deserialize": [Function],
    "input": {
      "ConditionExpression": undefined,
      "ExpressionAttributeNames": {
        "#nested": "nested",
        "#nestedBoolean": "nestedBoolean",
        "#nestedSet": "nestedSet",
        "#someBoolean": "someBoolean",
        "#someSet": "someSet",
      },
      "ExpressionAttributeValues": {
        ":addUpdateExpressionValue1": Set {
          "1",
          "2",
        },
        ":deleteUpdateExpressionValue2": Set {
          "a",
        },
        ":setUpdateExpressionValue0": true,
      },
      "Key": {
        "dataTimestamp": 2,
        "userId": "1",
      },
      "ReturnValues": undefined,
      "TableName": "myTable",
      "UpdateExpression": "SET #nested.#nestedBoolean = :setUpdateExpressionValue0 REMOVE #someBoolean ADD #someSet :addUpdateExpressionValue1 DELETE #nested.#nestedSet :deleteUpdateExpressionValue2",
    },
    "middlewareStack": {
      "add": [Function],
      "addRelativeTo": [Function],
      "applyToStack": [Function],
      "clone": [Function],
      "concat": [Function],
      "identify": [Function],
      "identifyOnResolve": [Function],
      "remove": [Function],
      "removeByTag": [Function],
      "resolve": [Function],
      "use": [Function],
    },
    "serialize": [Function],
  },
  "input": {
    "ConditionExpression": undefined,
    "ExpressionAttributeNames": {
      "#nested": "nested",
      "#nestedBoolean": "nestedBoolean",
      "#nestedSet": "nestedSet",
      "#someBoolean": "someBoolean",
      "#someSet": "someSet",
    },
    "ExpressionAttributeValues": {
      ":addUpdateExpressionValue1": Set {
        "1",
        "2",
      },
      ":deleteUpdateExpressionValue2": Set {
        "a",
      },
      ":setUpdateExpressionValue0": true,
    },
    "Key": {
      "dataTimestamp": 2,
      "userId": "1",
    },
    "ReturnValues": undefined,
    "TableName": "myTable",
    "UpdateExpression": "SET #nested.#nestedBoolean = :setUpdateExpressionValue0 REMOVE #someBoolean ADD #someSet :addUpdateExpressionValue1 DELETE #nested.#nestedSet :deleteUpdateExpressionValue2",
  },
  "inputKeyNodes": {
    "AttributeUpdates": {
      "*": {
        "Value": null,
      },
    },
    "Expected": {
      "*": {
        "AttributeValueList": [],
        "Value": null,
      },
    },
    "ExpressionAttributeValues": {},
    "Key": {},
  },
  "middlewareStack": {
    "add": [Function],
    "addRelativeTo": [Function],
    "applyToStack": [Function],
    "clone": [Function],
    "concat": [Function],
    "identify": [Function],
    "identifyOnResolve": [Function],
    "remove": [Function],
    "removeByTag": [Function],
    "resolve": [Function],
    "use": [Function],
  },
  "outputKeyNodes": {
    "Attributes": {},
    "ItemCollectionMetrics": {
      "ItemCollectionKey": {},
    },
  },
}
`;
